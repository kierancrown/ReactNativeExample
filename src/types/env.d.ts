declare module '@env' {
  export const DARKSKY_API_KEY: string;
  export const NEWS_API_KEY: string;
}
