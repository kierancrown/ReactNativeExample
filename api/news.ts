import { NEWS_API_KEY } from "@env"
import axios from "axios"
import { News } from "../types/news"

const getTopHeadlines = async (country: string): Promise<News.Article[] | undefined> => {
    try {
        const res = await axios.get(`https://newsapi.org/v2/top-headlines?country=${country}&apiKey=${NEWS_API_KEY}`)
        if (res.status === 200) {
            return Array.isArray(res.data.articles) ? res.data.articles as News.Article[] : []
        }
    } catch (error) {
        console.error("Error getting top headlines", error)
    }
}

export { getTopHeadlines }