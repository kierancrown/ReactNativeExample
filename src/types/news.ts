export module News {
  export interface Source {
    id: string;
    name: string;
  }

  export interface Article {
    source: Source;
    author: string;
    title: string;
    description: string;
    url: string;
    urlToImage: string;
    publishedAt: Date;
    content: string;
  }

  export interface HeadlinesResponse {
    status: string;
    totalResults: number;
    articles: Article[];
  }
}

export interface ParseResult {
  title: string | null;
  content: string | null;
  author: string | null;
  date_published: string | null;
  lead_image_url: string | null;
  dek: string | null;
  next_page_url: string | null;
  url: string;
  domain: string;
  excerpt: string | null;
  word_count: number;
  direction: 'ltr' | 'rtl';
  total_pages: number;
  rendered_pages: number;
}
