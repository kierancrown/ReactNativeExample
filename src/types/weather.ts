export module Weather {
  export interface Currently {
    time: number;
    summary: string;
    icon: string;
    nearestStormDistance: number;
    nearestStormBearing: number;
    precipIntensity: number;
    precipIntensityError: number;
    precipProbability: number;
    precipType: string;
    temperature: number;
    apparentTemperature: number;
    dewPoint: number;
    humidity: number;
    pressure: number;
    windSpeed: number;
    windGust: number;
    windBearing: number;
    cloudCover: number;
    uvIndex: number;
    visibility: number;
    ozone: number;
  }

  export interface Datum {
    time: number;
    precipIntensity: number;
    precipIntensityError: number;
    precipProbability: number;
    precipType: string;
  }

  export interface Minutely {
    summary: string;
    icon: string;
    data: Datum[];
  }

  export interface Datum2 {
    time: number;
    summary: string;
    icon: string;
    precipIntensity: number;
    precipProbability: number;
    precipType: string;
    temperature: number;
    apparentTemperature: number;
    dewPoint: number;
    humidity: number;
    pressure: number;
    windSpeed: number;
    windGust: number;
    windBearing: number;
    cloudCover: number;
    uvIndex: number;
    visibility: number;
    ozone: number;
  }

  export interface Hourly {
    summary: string;
    icon: string;
    data: Datum2[];
  }

  export interface Datum3 {
    time: number;
    summary: string;
    icon: string;
    sunriseTime: number;
    sunsetTime: number;
    moonPhase: number;
    precipIntensity: number;
    precipIntensityMax: number;
    precipIntensityMaxTime: number;
    precipProbability: number;
    precipType: string;
    temperatureHigh: number;
    temperatureHighTime: number;
    temperatureLow: number;
    temperatureLowTime: number;
    apparentTemperatureHigh: number;
    apparentTemperatureHighTime: number;
    apparentTemperatureLow: number;
    apparentTemperatureLowTime: number;
    dewPoint: number;
    humidity: number;
    pressure: number;
    windSpeed: number;
    windGust: number;
    windGustTime: number;
    windBearing: number;
    cloudCover: number;
    uvIndex: number;
    uvIndexTime: number;
    visibility: number;
    ozone: number;
    temperatureMin: number;
    temperatureMinTime: number;
    temperatureMax: number;
    temperatureMaxTime: number;
    apparentTemperatureMin: number;
    apparentTemperatureMinTime: number;
    apparentTemperatureMax: number;
    apparentTemperatureMaxTime: number;
  }

  export interface Daily {
    summary: string;
    icon: string;
    data: Datum3[];
  }

  export interface Alert {
    title: string;
    regions: string[];
    severity: string;
    time: number;
    expires: number;
    description: string;
    uri: string;
  }

  export interface Flags {
    sources: string[];
    'nearest-station': number;
    units: string;
  }

  export interface RootObject {
    latitude: number;
    longitude: number;
    timezone: string;
    currently: Weather.Currently;
    minutely: Weather.Minutely;
    hourly: Weather.Hourly;
    daily: Weather.Daily;
    alerts: Weather.Alert[];
    flags: Weather.Flags;
    offset: number;
  }
}
