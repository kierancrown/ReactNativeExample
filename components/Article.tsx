import React from 'react';
import {
    Image,
    StyleSheet,
    Text,
    useColorScheme,
    View,
} from 'react-native';
import { News } from '../types/news';
import moment from "moment";

const ArticleCard = ({ urlToImage, source, title, publishedAt }: News.Article) => {
    const isDarkMode = useColorScheme() === 'dark';

    return (
        <View style={{ backgroundColor: isDarkMode ? '#000' : '#fff', ...largeStyles.articleCard }}>
            <Image
                fadeDuration={250}
                source={urlToImage == null ? require('../assets/placeholder.png') : { uri: urlToImage }}
                defaultSource={require('../assets/placeholder.png')}
                resizeMode='cover'
                style={largeStyles.articleMedia} />
            <View style={largeStyles.textWrapper}>
                <Text style={{ color: isDarkMode ? '#fff' : '#000', ...largeStyles.articleTitle }}>{title}</Text>
                <View style={largeStyles.metadata}>
                    <Text style={{ color: isDarkMode ? '#fff' : '#999', ...largeStyles.sourceName }}>{source.name.toUpperCase()}</Text>
                    <Text style={{ color: isDarkMode ? '#fff' : '#999', ...largeStyles.timeAgo }}>{moment(publishedAt).fromNow().toUpperCase()}</Text>
                </View>
            </View>
        </View >
    );
};

const largeStyles = StyleSheet.create({
    articleCard: {
        justifyContent: 'center',
        alignItems: 'center',
        margin: 10,
        borderRadius: 10,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.23,
        shadowRadius: 2.62,
        elevation: 4,
    },
    articleMedia: {
        flex: 4,
        width: '100%',
        height: undefined,
        aspectRatio: 1.5,
        borderTopLeftRadius: 10,
        borderTopRightRadius: 10,
    },
    metadata: {
        justifyContent: 'center',
        flexDirection: 'row',
        opacity: .75,
        width: '100%',
    },
    textWrapper: {
        marginTop: 5,
        paddingHorizontal: 8,
        paddingVertical: 4,
        flex: 1,
        marginRight: 'auto',
    },
    sourceName: {
        fontWeight: '600',
        fontSize: 10,
        marginVertical: 5,
    },
    timeAgo: {
        fontWeight: '600',
        fontSize: 10,
        marginVertical: 5,
        marginLeft: 'auto'
    },
    articleTitle: {
        fontWeight: '600',
        fontSize: 18,
        marginVertical: 5,
    }
})

export default ArticleCard;