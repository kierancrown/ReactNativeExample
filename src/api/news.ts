import {NEWS_API_KEY} from '@env';
import axios from 'axios';
import {News} from '../types/news';

const getTopHeadlines = async (
  country: string,
): Promise<News.Article[] | undefined> => {
  return new Promise(async (resolve, reject) => {
    try {
      const res = await axios.get(
        `https://newsapi.org/v2/top-headlines?country=${country}&apiKey=${NEWS_API_KEY}`,
      );
      if (res.status === 200) {
        resolve(
          Array.isArray(res.data.articles)
            ? (res.data.articles as News.Article[])
            : [],
        );
      }
    } catch (error) {
      reject(error);
    }
  });
};

export {getTopHeadlines};
