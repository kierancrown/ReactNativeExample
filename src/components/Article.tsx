import React from 'react';
import {Image, StyleSheet, Text, useColorScheme, View} from 'react-native';
import {News} from '../types/news';
import moment from 'moment';

const ArticleCard = ({
  urlToImage,
  source,
  title,
  publishedAt,
}: News.Article) => {
  const isDarkMode = useColorScheme() === 'dark';

  const styles = StyleSheet.create({
    card: {
      backgroundColor: isDarkMode ? '#000' : '#fff',
      justifyContent: 'center',
      alignItems: 'center',
      margin: 10,
      borderRadius: 10,
      shadowColor: '#000',
      shadowOffset: {
        width: 0,
        height: 2,
      },
      shadowOpacity: 0.23,
      shadowRadius: 2.62,
      elevation: 4,
    },
    media: {
      flex: 4,
      width: '100%',
      height: undefined,
      aspectRatio: 1.5,
      borderTopLeftRadius: 10,
      borderTopRightRadius: 10,
    },
    metadata: {
      justifyContent: 'center',
      flexDirection: 'row',
      opacity: 0.75,
      width: '100%',
    },
    textWrapper: {
      marginTop: 5,
      paddingHorizontal: 8,
      paddingVertical: 4,
      flex: 1,
      marginRight: 'auto',
    },
    sourceName: {
      color: isDarkMode ? '#fff' : '#999',
      fontWeight: '600',
      fontSize: 10,
      marginVertical: 5,
    },
    timeAgo: {
      color: isDarkMode ? '#fff' : '#999',
      fontWeight: '600',
      fontSize: 10,
      marginVertical: 5,
      marginLeft: 'auto',
    },
    title: {
      color: isDarkMode ? '#fff' : '#000',
      fontWeight: '600',
      fontSize: 18,
      marginVertical: 5,
    },
  });

  return (
    <View style={styles.card}>
      <Image
        fadeDuration={250}
        source={
          urlToImage == null
            ? require('../assets/placeholder.png')
            : {uri: urlToImage}
        }
        defaultSource={require('../assets/placeholder.png')}
        resizeMode="cover"
        style={styles.media}
      />
      <View style={styles.textWrapper}>
        <Text style={styles.title}>{title}</Text>
        <View style={styles.metadata}>
          <Text style={styles.sourceName}>{source.name.toUpperCase()}</Text>
          <Text style={styles.timeAgo}>
            {moment(publishedAt).fromNow().toUpperCase()}
          </Text>
        </View>
      </View>
    </View>
  );
};

export default ArticleCard;
