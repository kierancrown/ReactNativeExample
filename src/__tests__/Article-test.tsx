import * as React from 'react';
import Article from '../components/Article';
import renderer from 'react-test-renderer';
import {News} from '../types/news';

const mockProps: News.Article = {
  urlToImage: 'https://www.example.com/image.jpg',
  source: {
    id: '1',
    name: 'Example',
  },
  title: 'Example title',
  author: 'Example author',
  description: 'Example description',
  url: 'https://www.example.com',
  content: 'Example content',
  publishedAt: new Date(),
};

it('renders correctly', () => {
  const tree = renderer.create(<Article {...mockProps} />);
  expect(tree).toMatchSnapshot();
});
