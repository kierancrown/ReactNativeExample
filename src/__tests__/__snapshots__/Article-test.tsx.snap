// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#eee",
      "borderRadius": 10,
      "elevation": 4,
      "justifyContent": "center",
      "marginVertical": 10,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.23,
      "shadowRadius": 2.62,
    }
  }
>
  <Image
    defaultSource={
      Object {
        "testUri": "../../../src/assets/placeholder.png",
      }
    }
    fadeDuration={250}
    resizeMode="cover"
    source={
      Object {
        "uri": "https://www.example.com/image.jpg",
      }
    }
    style={
      Object {
        "aspectRatio": 1.5,
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "flex": 4,
        "height": undefined,
        "width": "100%",
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "marginRight": "auto",
        "marginTop": 5,
        "paddingHorizontal": 8,
        "paddingVertical": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000",
          "fontSize": 18,
          "fontWeight": "600",
          "marginVertical": 5,
        }
      }
    >
      Example title
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 0.75,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#999",
            "fontSize": 10,
            "fontWeight": "600",
            "marginVertical": 5,
          }
        }
      >
        EXAMPLE
      </Text>
      <Text
        style={
          Object {
            "color": "#999",
            "fontSize": 10,
            "fontWeight": "600",
            "marginLeft": "auto",
            "marginVertical": 5,
          }
        }
      >
        A FEW SECONDS AGO
      </Text>
    </View>
  </View>
</View>
`;
